equinox/nn/attention.py

```diff
@@ -11,6 +11,41 @@
 from .linear import Linear
 
 
+def dot_product_attention_weights(
+    query: Array["query_seq_length", "qk_size"],  # noqa: F821
+    key: Array["kv_seq_length", "qk_size"],  # noqa: F821
+    mask: Optional[Array["query_seq_length", "kv_seq_length"]] = None,  # noqa: F821
+) -> Array["query_seq_length", "kv_seq_length"]:  # noqa: F821
+
+    logits = jnp.einsum("sd,Sd->sS", query, key)
+    logits = logits / math.sqrt(query.shape[-1])
+    if mask is not None:
+        if mask.shape != logits.shape:
+            raise ValueError(
+                f"mask must have shape (query_seq_length, "
+                f"kv_seq_length)=({query.shape[0]}, "
+                f"{key.shape[0]}). Got {mask.shape}."
+            )
+        logits = jnp.where(mask, logits, -jnp.inf)
+
+    return jax.nn.softmax(logits, axis=-1)
+
+
+def dot_product_attention(
+    query: Array["query_seq_length", "qk_size"],  # noqa: F821
+    key: Array["kv_seq_length", "qk_size"],  # noqa: F821
+    value: Array["kv_seq_length", "value_size"],  # noqa: F821
+    mask: Optional[Array["query_seq_length", "kv_seq_length"]] = None,  # noqa: F821
+    dropout_fn: Optional[Callable[[Array], Array]] = None,
+) -> Array["query_seq_length", "value_size"]:  # noqa: F821
+
+    weights = dot_product_attention_weights(query, key, mask)
+    if dropout_fn:
```

Comment:
Can this be `if dropout_fn is not None`? (I'm not a fan of implicit-cast-to-bool.)